src/main.cpp

```diff
@@ -2968,6 +3043,95 @@ bool static LoadBlockIndexDB()
     return true;
 }
 
+bool BlockFileIsOpenable(int nFile)
+{
+    CDiskBlockPos pos(nFile, 0);
+    if (!vinfoBlockFile[nFile].nSize)
+        return false;
+    if (CAutoFile(OpenBlockFile(pos, true), SER_DISK, CLIENT_VERSION).IsNull()) {
+        ClearBlockFileInfo(nFile);
+        return false;
+    }
+    return true;
+}
+
+bool UndoFileIsOpenable(int nFile)
+{
+    CDiskBlockPos pos(nFile, 0);
+    if (!vinfoBlockFile[nFile].nUndoSize)
+        return false;
+    if (CAutoFile(OpenUndoFile(pos, true), SER_DISK, CLIENT_VERSION).IsNull()) {
+        ClearUndoFileInfo(nFile);
+        return false;
+    }
+    return true;
+}
+
+bool DataFilesAreOpenable(int nFile)
+{
+    if (BlockFileIsOpenable(nFile) && UndoFileIsOpenable(nFile))
+        return true;
+    return false;
+}
+
+int LastBlockInFile(int nFile)
+{
+    return vinfoBlockFile[nFile].nHeightLast;
+}
+
+
+bool CheckBlockFiles()
+{
+    LOCK(cs_main);
+    // Check presence of blk files
+    int nKeepMinBlksFromHeight = nPrune ? (max((int)(chainActive.Height() - MIN_BLOCKS_TO_KEEP), 0)) : 0;
+    LogPrintf("Checking all required data for active chain is available (mandatory from height %i to %i)\n", nKeepMinBlksFromHeight, max(chainActive.Height(), 0));
+    set<int> setDataPruned, setUndoPruned;
+    setDataFilePrunable.clear();
+    setUndoFilePrunable.clear();
+    for (CBlockIndex* pindex = chainActive.Tip(); pindex && pindex->pprev; pindex = pindex->pprev) {
```

Comment:
This will not iterate over side branches of the block chain, and may perhaps miss things that way. The chances for that are really low, as this code still checks the last height in each affected file, and then deletes whole files, so unless a file consists solely of side branch blocks, it will still be processed.

However, I think this flaw indicates a conceptual mistake: this function should iterate over files, rather than over blocks. That would be faster too. The concistency check can be done on blocks separately, but only needs to be done for the last blocks I guess?